Subtests with the same name conflict on -update

When multiple top-level test cases contain subtests, and any of those subtests share the same name (across different parents), a number of issues arise.

In some cases, the golden file for a different test is used in the comparison, and is reported as a mismatch (since its content typically differs). In other cases, the golden file for a test appears to be not present at all (even though it was present prior to the run), and is treated as the first run of the test (creating the golden file, and failing the test). In yet another case, the rename fails and reports the file as not existing.

This appears to be a result of the golden files being copied to the temp dir, without the parent test being part of the file path. As such, they are copied directly under `testdata/<name>.golden`, where (sub)tests with the same name would conflict. I suspect this would also impact top-level tests where their name conflicts with a subtest e.g. `TestExample` and `TestParent/TestExample` (though this naming is unusual).

Example test case:

```
func TestEqual_subtestSameNames1(t *testing.T) {
	testEqualSubtestSameNames(t)
}

func TestEqual_subtestSameNames2(t *testing.T) {
	testEqualSubtestSameNames(t)
}

func testEqualSubtestSameNames(t *testing.T) {
	t.Parallel()

	parent := t.Name()

	testTable := []string{
		"first",
		"second",
		"third",
	}

	for _, name := range testTable {
		name := name

		t.Run(name, func(t *testing.T) {
			// Subtests are intentionally not run in parallel, as that makes this issue more easily reproducible

			Equal(t, fmt.Sprintf("%s :: %s", parent, name))
		})
	}
}
```

Failure case 1:

```
--- FAIL: TestEqual_subtestSameNames1 (0.01s)
    --- FAIL: TestEqual_subtestSameNames1/second (0.01s)
        autogold.go:136: mismatch (-want +got):
            -- want
            ++ got
            @@ -1 +1 @@
            "TestEqual_subtestSameNames2 :: second"
            "TestEqual_subtestSameNames1 :: second"
            
    --- FAIL: TestEqual_subtestSameNames1/third (0.00s)
        autogold.go:136: mismatch (-want +got):
            -- want
            ++ got
            @@ -1 +1 @@
            "TestEqual_subtestSameNames2 :: third"
            "TestEqual_subtestSameNames1 :: third"
            
--- FAIL: TestEqual_subtestSameNames2 (0.01s)
    --- FAIL: TestEqual_subtestSameNames2/second (0.00s)
        autogold.go:136: mismatch (-want +got):
            -- want
            ++ got
            @@ -1 +1 @@
            "TestEqual_subtestSameNames2 :: second"
            
    --- FAIL: TestEqual_subtestSameNames2/third (0.00s)
        autogold.go:136: mismatch (-want +got):
            -- want
            ++ got
            @@ -1 +1 @@
            "TestEqual_subtestSameNames2 :: third"
            
FAIL
FAIL    github.com/hexops/autogold      0.765s
ok      github.com/hexops/autogold/example      0.850s
?       github.com/hexops/autogold/internal/test        [no test files]
FAIL
```

Failure case 2:

```
--- FAIL: TestWant_parallel1 (0.00s)
    --- FAIL: TestWant_parallel1/1-first (0.00s)
        autogold.go:92: rename testdata/TestWant_parallel1/1-first.golden /var/folders/2b/5543qvds4msgy2rbyldbrs0c0000gq/T/go-golden-1664517986/1-first.golden: no such file or directory
--- FAIL: TestWant_parallel2 (0.00s)
    --- FAIL: TestWant_parallel2/2-first (0.00s)
        autogold.go:92: rename testdata/TestWant_parallel2/2-first.golden /var/folders/2b/5543qvds4msgy2rbyldbrs0c0000gq/T/go-golden-1664517986/2-first.golden: no such file or directory
--- FAIL: TestEqual_subtestSameNames1 (0.01s)
    --- FAIL: TestEqual_subtestSameNames1/first (0.00s)
        autogold.go:92: rename testdata/TestEqual_subtestSameNames1/first.golden /var/folders/2b/5543qvds4msgy2rbyldbrs0c0000gq/T/go-golden-1664517986/first.golden: no such file or directory
--- FAIL: TestEqual_subtestSameNames2 (0.01s)
    --- FAIL: TestEqual_subtestSameNames2/first (0.01s)
        autogold.go:92: rename testdata/TestEqual_subtestSameNames2/first.golden /var/folders/2b/5543qvds4msgy2rbyldbrs0c0000gq/T/go-golden-1664517986/first.golden: no such file or directory
FAIL
FAIL    github.com/hexops/autogold      0.420s
ok      github.com/hexops/autogold/example      0.402s
?       github.com/hexops/autogold/internal/test        [no test files]
FAIL
```

--- a/.github/workflows/go-ci.yml
+++ b/.github/workflows/go-ci.yml
@@ -14,7 +14,8 @@ jobs:
       - run: go test ./...
       # Testing with -update enabled isn't intended to update the golden files, but to help validate that there aren't
       # problems when updating is enabled, race conditions in the update process etc
-      - run: go test -race ./... -update
+      # Tests are run multiple times to help expose race-conditions as they aren't always reproduced on a single run
+      - run: go test -race -count 5 ./... -update
       - run: go test -race -coverprofile=coverage.txt -covermode=atomic ./...
       - name: Upload code coverage
         uses: codecov/codecov-action@v1

--- a/autogold.go
+++ b/autogold.go
@@ -76,6 +76,10 @@ func Equal(t *testing.T, got interface{}, opts ...Option) {
 			t.Fatal(err)
 		}
 		grabLock()
+
+		// cleanDir may not be set until mkTempDir(), so we can't assign this earlier
+		tmpdir := filepath.Join(cleanDir, dir)
+
 		_, ok := cleaned[dir]
 		if !ok {
 			// Move all .golden files in the directory into the temp dir.
@@ -85,8 +89,14 @@ func Equal(t *testing.T, got interface{}, opts ...Option) {
 				cleanMu.Unlock()
 				t.Fatal(err)
 			}
+
+			if err := os.MkdirAll(tmpdir, 0o700); err != nil {
+				cleanMu.Unlock()
+				t.Fatal(err)
+			}
+
 			for _, match := range matches {
-				err := os.Rename(match, filepath.Join(cleanDir, filepath.Base(match)))
+				err := os.Rename(match, filepath.Join(tmpdir, filepath.Base(match)))
 				if err != nil {
 					cleanMu.Unlock()
 					t.Fatal(err)
@@ -97,7 +107,7 @@ func Equal(t *testing.T, got interface{}, opts ...Option) {
 		cleanMu.Unlock()
 
 		// Move the golden file for this test back into the testdata dir, if it exists.
-		tmpFile := filepath.Join(cleanDir, filepath.Base(fileName+".golden"))
+		tmpFile := filepath.Join(tmpdir, filepath.Base(fileName+".golden"))
 		err := os.Rename(tmpFile, outFile)
 		if err != nil && !os.IsNotExist(err) {
 			t.Fatal(err)

--- a/testdata/TestEqual_subtestSameNames1/first.golden
+++ b/testdata/TestEqual_subtestSameNames1/first.golden
@@ -1,0 +1,1 @@
+"TestEqual_subtestSameNames1 :: first"

--- a/testdata/TestEqual_subtestSameNames1/second.golden
+++ b/testdata/TestEqual_subtestSameNames1/second.golden
@@ -1,0 +1,1 @@
+"TestEqual_subtestSameNames1 :: second"

--- a/testdata/TestEqual_subtestSameNames1/third.golden
+++ b/testdata/TestEqual_subtestSameNames1/third.golden
@@ -1,0 +1,1 @@
+"TestEqual_subtestSameNames1 :: third"

--- a/testdata/TestEqual_subtestSameNames2/first.golden
+++ b/testdata/TestEqual_subtestSameNames2/first.golden
@@ -1,0 +1,1 @@
+"TestEqual_subtestSameNames2 :: first"

--- a/testdata/TestEqual_subtestSameNames2/second.golden
+++ b/testdata/TestEqual_subtestSameNames2/second.golden
@@ -1,0 +1,1 @@
+"TestEqual_subtestSameNames2 :: second"

--- a/testdata/TestEqual_subtestSameNames2/third.golden
+++ b/testdata/TestEqual_subtestSameNames2/third.golden
@@ -1,0 +1,1 @@
+"TestEqual_subtestSameNames2 :: third"

--- a/want_test.go
+++ b/want_test.go
@@ -108,6 +108,36 @@ func Test_getPackageNameAndPath_subdir(t *testing.T) {
 	}
 }
 
+func TestEqual_subtestSameNames1(t *testing.T) {
+	testEqualSubtestSameNames(t)
+}
+
+func TestEqual_subtestSameNames2(t *testing.T) {
+	testEqualSubtestSameNames(t)
+}
+
+func testEqualSubtestSameNames(t *testing.T) {
+	t.Parallel()
+
+	parent := t.Name()
+
+	testTable := []string{
+		"first",
+		"second",
+		"third",
+	}
+
+	for _, name := range testTable {
+		name := name
+
+		t.Run(name, func(t *testing.T) {
+			// Subtests are intentionally not run in parallel, as that makes this issue more easily reproducible
+
+			Equal(t, fmt.Sprintf("%s :: %s", parent, name))
+		})
+	}
+}
+
 func Benchmark_getPackageNameAndPath_cached(b *testing.B) {
 	// Wipe the cache, as it was populated by other tests.
 	getPackageNameAndPathCacheMu.Lock()

